README.md
# Polyglot code offline layout

This is a quite hacky script to calculate Voronoi Treemap layouts for JSON files produced by the polyglot-code-scanner program.

For an overview see <https://polyglot.korny.info>

For more detailed instructions on running this tool see <https://polyglot.korny.info/tools/layout/howto>

You can run this script using node.js, or via Docker - or binary builds are coming

## WORK IN PROGRESS WARNING

I'm doing a lot of changes right now - if you fetch the current code, things may break.

Especially note, I'm changed the data file formats created by the explorer and used by the scanner - I've added version number checks, but data files from the Scanner must match expectations of the Explorer, so for now it's a bit of "make sure you pull changes often" or things will break.


## Running this script using node.js

1. Install node.js - see https://nodejs.org/en/download/package-manager/
2. clone the code from https://github.com/kornysietsma/polyglot-code-offline-layout
3. In the checked out repository, run `npm install` to fetch all the dependencies

Then you can run it with

```sh
$ node layout.js -i input.json -o output.json
```

If you have multiple repositories in your source json, e.g. for a group of projects or microservices, you can ask for a pretty circle-packed algorithm for the main diagram with `-c`:

```sh
$ node layout.js -c -i input.json -o output.json
```

The `-i` and `-o` parameters are optional, if you don't supply them the script will read stdin and send output to stdout, so you can use unix-style pipes, or :

```sh
$ node layout.js < input.json > output.json
```

There are some optional parameters:

- -h for help
- -c to use circle packing for the top level shaping, voronoi for the rest
- -g (the default) to accept an imperfect voronoi map - if you turn this off, the script will fail if it can't find a precise layout
- -p NNN - specify how many points to use to draw the main circles around voronoi treemaps.  The default is 128, which is a pretty smooth circle.  You could specify `-p 6` and it will use hexagons!

## Binary executables

If you don't want to install node.js, you can fetch a bundled executable file instead.

I am using [pkg](https://www.npmjs.com/package/pkg) to package up runnable executables.

### Fetching binary releases

Binary releases are published to <https://github.com/kornysietsma/polyglot-code-offline-layout/releases>

You should be able to download the right file for your platform here.  Note I haven't tested this at all on Windows yet!  And I only have a single generic build for linux and OSX - if you are on another platform, you'll have to use node.js for now.

### Getting binaries to work on mac OSX

OSX has a system to protect you from malware, called 'gatekeeper' - by default on OSX if you try to run the binary app, you'll get an error "app is from an unknown developer".  To bypass this I'd have to go through a complex extra stage of signing and bundling my tool as an application, which I'm avoiding for now.

If you are happy that the binary is safe - all the code that builds it is on github, so you should be able to inspect it yourself - you can run it by stripping the extra attributes that OSX adds when you download it:

```sh
unzip polyglot-code-offline-layout-macos.zip
xattr -d com.apple.quarantine polyglot-code-offline-layout
```

Then you can move the `polyglot-code-offline-layout` binary to wherever you want.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [0.3.1] - 2022-11-15

### Changed

- minor but important change - make sure exit status is set on errors, for when this is called by scripts

## [0.3.0] - 2022-10-28

### Changed

- using new data file based on v1.0.4 data file format - backward incompatible change!

## [0.2.0] - 2020-08-01

### Changed

- using new data formats based on v1.0.0 data file format


## [0.1.3] - 2020-11-17

### Changed

- using d3-voronoi-map official release 2.1.0 as my fork is no longer needed due to upstream bug fixes

## [0.1.2] - 2020-09-01

### Added

- this changelog!

### Changed

- using antonyurchenko/git-release action for smoother releases

## [0.1.1] - 2020-08-29

### Added

- Using github actions to publish releases
- First packaged release with binaries


offline-voronoi.js
import { promises as fs, existsSync } from 'fs';
import yargs from 'yargs';
import dvm from 'd3-voronoi-map';

const voronoiMapSimulation = dvm.voronoiMapSimulation;

const debug = false;

import d3 from 'd3';
// import vtm from 'd3-voronoi-treemap';

function computeCirclingPolygon(points, radius) {
  const increment = (2 * Math.PI) / points;
  const circlingPolygon = [];

  for (let a = 0, i = 0; i < points; i++, a += increment) {
    circlingPolygon.push([radius * Math.cos(a), radius * Math.sin(a)]);
  }

  return circlingPolygon;
}

function flareWeightLoc(d) {
  if (d.data === undefined) return 0;
  if (d.data.loc === undefined) return 0;
  return d.data.loc.code;
}

function pruneWeightlessNodes(hierarchy) {
  if (hierarchy.children !== undefined) {
    // eslint-disable-next-line no-param-reassign
    hierarchy.children = hierarchy.children.filter((node) => node.value > 0);
    hierarchy.children.forEach((child) => pruneWeightlessNodes(child));
  }
}

function addPaths(pathSoFar, node) {
  let path;
  if (pathSoFar === null) {
    path = ''; // not 'flare' - could use '/' or null - but this is nicer for output
  } else {
    if (pathSoFar === '') {
      path = node.name;
    } else {
      path = `${pathSoFar}/${node.name}`;
    }
  }
  const children = node.children
    ? node.children.map((n) => addPaths(path, n))
    : undefined;
  return {
    name: node.name,
    path,
    children: children,
    layout: node.layout,
    value: node.value,
    data: node.data,
  };
}

function calculate_values(node) {
  if (node.children) {
    for (const n of node.children) {
      calculate_values(n);
    }
    const tot = node.children.map((n) => n.value).reduce((a, b) => a + b, 0);
    node.value = tot;
  } else {
    node.value = flareWeightLoc(node);
  }
}

function calculateVoronoi(nameSoFar, node, clipPolygon, center) {
  const name = nameSoFar ? `${nameSoFar}/${node.name}` : node.name;
  node.layout = {
    polygon: clipPolygon,
    center,
    algorithm: 'voronoi',
  };

  if (!node.children) {
    return;
  }
  if (debug) {
    console.log(
      `calculating map for ${name} with ${node.children.length} children and a clip polygon with ${clipPolygon.length} vertices`
    );
  }

  const MY_MAX_COUNT = 500;
  let metaCount = 0;
  let metaEnded = false;
  const META_MAX_COUNT = 100;
  while (!metaEnded) {
    try {
      var simulation = voronoiMapSimulation(node.children)
        .maxIterationCount(MY_MAX_COUNT)
        .minWeightRatio(0.01)
        .weight((d) => d.value)
        .clip(clipPolygon)
        .stop();

      var state = simulation.state();

      let tickCount = 0;
      let warningTime = Date.now();
      while (!state.ended) {
        tickCount += 1;
        const now = Date.now();
        if (now - warningTime > 10000) {
          // every 10 seconds
          warningTime = now;
          console.log(
            `slow voronoi processing of ${name} with ${node.children.length} children, tick count: ${tickCount}`
          );
        }
        simulation.tick();
        state = simulation.state();
      }
      if (tickCount === MY_MAX_COUNT) {
        if (metaCount < META_MAX_COUNT) {
          metaCount = metaCount + 1;
          console.log(
            `processing ${name} with ${node.children.length} children - Exceeded tick count ${tickCount} - retrying from scratch, try ${metaCount}`
          );
        } else {
          console.error('Too many meta retries - stopping anyway');
          metaEnded = true;
        }
      } else {
        metaEnded = true;
      }
    } catch (e) {
      // re-try from scratch but only after predictable exceptions
      console.error('caught e', e);
      if (!e instanceof Error) {
        console.error('not Error');
        throw e;
      }
      if (e.message == 'bad_polygons' || e.message == 'overweight_loop') {
        metaCount = metaCount + 1;
        if (metaCount < META_MAX_COUNT) {
          console.error(`caught ${e.message}, retrying`, metaCount);
        } else {
          console.error(`caught ${e.message}, too many errors!`, metaCount);
          metaEnded = true; // but fail anyway
          throw e;
        }
      } else {
        console.error(`unhandled exception ${e.message} - rethrowing`);
        throw e;
      }
    }
  }
  var polygons = state.polygons;

  for (const polygon of polygons) {
    const pdata = polygon.map((d) => d);
    calculateVoronoi(
      name,
      polygon.site.originalObject.data.originalData,
      pdata,
      [polygon.site.x, polygon.site.y]
    );
  }
}

async function main({ input, output, points, circles }) {
  const rawData = await fs.readFile(input, 'utf-8');
  const width = 1024;
  const parsedData = JSON.parse(rawData);
  //   console.log(parsedData);
  console.log('getting values recursively');
  calculate_values(parsedData);
  console.log('pruning empty nodes');
  pruneWeightlessNodes(parsedData);

  // top level clip shape
  if (circles) {
    // area = pi r^2 so r = sqrt(area/pi) or just use sqrt(area) for simplicity
    console.log('getting children');
    const children = parsedData.children.map((child) => {
      console.log('child weight', child.value);
      return { r: Math.sqrt(child.value), originalObject: child };
    });
    d3.packSiblings(children);
    console.log('children', children);
    // top level layout
    const enclosingCirle = d3.packEnclose(children);
    const { x, y, r } = enclosingCirle;
    // TODO: offset by x/y
    console.log('enclosing circle radius', r);
    parsedData.layout = {
      polygon: computeCirclingPolygon(points, r),
      center: [0, 0],
      width: r * 2,
      height: r * 2,
      algorithm: 'circlePack',
    };
    console.log('top layout', parsedData.layout);

    for (const child of children) {
      const clipPolygon = computeCirclingPolygon(
        points,
        child.r
      ).map(([x, y]) => [x + child.x, y + child.y]);
      const center = [child.x, child.y];
      console.log('calculating voronoi for ', child.originalObject.name);
      calculateVoronoi(
        child.originalObject.name,
        child.originalObject,
        clipPolygon,
        center
      );
      child.originalObject.layout.width = child.r;
      child.originalObject.layout.height = child.r;
    }
  } else {
    const clipPolygon = computeCirclingPolygon(points, width / 2);
    const center = [0, 0];

    calculateVoronoi(null, parsedData, clipPolygon, center);

    parsedData.layout.width = width;
    parsedData.layout.height = width;
  }

  const results = addPaths(null, parsedData);

  console.log('saving');
  await fs.writeFile(output, JSON.stringify(results));
  return 'OK';
}

const argv = yargs
  //   .usage('$0 -i [input]', 'add voronoi data to json input file')
  .usage('$0 [options]')
  .alias('i', 'input')
  .describe('i', 'Input JSON file')
  .alias('o', 'output')
  .describe('o', 'Output JSON file')
  .alias('p', 'points')
  .number('p')
  .default('p', 128)
  .describe('p', 'number of points in the initial bounding circle/polygon')
  .describe('c', 'use circle packing for top level')
  .boolean('c')
  .alias('c', 'circles')
  .default('c', false)
  .demandOption(['i', 'o'])
  .help('h')
  .alias('h', 'help')
  .check((argv) => {
    if (existsSync(argv.i)) {
      return true;
    }
    throw new Error(`Argument check failed: ${argv.i} does not exist`);
  }).argv;

const args = {
  input: argv.input,
  output: argv.output,
  points: argv.points,
  circles: argv.circles,
};

main(args).then(
  (result) => {
    console.error('done.', result);
  },
  (err) => {
    console.error('Error thrown!', err);
  }
);


repo_change_stats.js
import { promises as fs, existsSync } from 'fs';
import yargs from 'yargs';
import moment from 'moment';

function accumulateDetails(yearMap, details) {
  details.forEach((detail) => {
    const year = moment.unix(detail.commit_day).year();
    const yearData = yearMap.get(year) || { commits: 0, lines: 0 };
    yearData.commits += detail.commits;
    yearData.lines += detail.lines_added + detail.lines_deleted;
    yearMap.set(year, yearData);
  });
}

function accumulateYearData(node, yearMap) {
  if (node.data && node.data.git && node.data.git.details) {
    accumulateDetails(yearMap, node.data.git.details);
  }
  if (node.children) {
    node.children.forEach((child) => {
      accumulateYearData(child, yearMap);
    });
  }
}

function yearData(repoName, node) {
  const yearMap = new Map();
  accumulateYearData(node, yearMap);
  return yearMap;
}

async function main({ input, output, points, circles }) {
  const rawData = await fs.readFile(input, 'utf-8');
  const parsedData = JSON.parse(rawData);
  // output is CSV for easy fiddling in excel
  console.log('repo, year, commits, lines');
  parsedData.children.forEach((child) => {
    if (child.children) {
      const repoName = child.name;
      const data = yearData(repoName, child);
      [...data.keys()].forEach((year) => {
        const yearData = data.get(year);
        console.log(
          `"${repoName}",${year},${yearData.commits},${yearData.lines}`
        );
      });
    }
  });
}

const argv = yargs
  //   .usage('$0 -i [input]', 'add voronoi data to json input file')
  .usage('$0 [options]')
  .alias('i', 'input')
  .describe('i', 'Input JSON file')
  .help('h')
  .alias('h', 'help')
  .check((argv) => {
    if (existsSync(argv.i)) {
      return true;
    }
    throw new Error(`Argument check failed: ${argv.i} does not exist`);
  }).argv;

const args = {
  input: argv.input,
};

main(args).then(
  (result) => {
    console.error('done.', result);
  },
  (err) => {
    console.error('Error thrown!', err);
  }
);


layout.js
#!/usr/bin/env node

const { promises: fs, existsSync } = require('fs');
const yargs = require('yargs');
const dvm = require('d3-voronoi-map');
const d3 = require('d3');

const voronoiMapSimulation = dvm.voronoiMapSimulation;

const debug = false;

// import vtm from 'd3-voronoi-treemap';

function computeCirclingPolygon(points, radius) {
  const increment = (2 * Math.PI) / points;
  const circlingPolygon = [];

  for (let a = 0, i = 0; i < points; i++, a += increment) {
    circlingPolygon.push([radius * Math.cos(a), radius * Math.sin(a)]);
  }

  return circlingPolygon;
}

function flareWeightLoc(d) {
  if (d.data === undefined) return 0;
  if (d.data.loc === undefined) return 0;
  return d.data.loc.code;
}

function pruneWeightlessNodes(hierarchy) {
  if (hierarchy.children !== undefined) {
    // eslint-disable-next-line no-param-reassign
    hierarchy.children = hierarchy.children.filter((node) => node.value > 0);
    hierarchy.children.forEach((child) => pruneWeightlessNodes(child));
  }
}

function addPaths(pathSoFar, node) {
  let path;
  if (pathSoFar === null) {
    path = ''; // not 'flare' - could use '/' or null - but this is nicer for output
  } else {
    if (pathSoFar === '') {
      path = node.name;
    } else {
      path = `${pathSoFar}/${node.name}`;
    }
  }
  const children = node.children
    ? node.children.map((n) => addPaths(path, n))
    : undefined;
  return {
    name: node.name,
    path,
    children: children,
    layout: node.layout,
    value: node.value,
    data: node.data,
  };
}

function calculate_values(node) {
  if (node.children) {
    for (const n of node.children) {
      calculate_values(n);
    }
    const tot = node.children.map((n) => n.value).reduce((a, b) => a + b, 0);
    node.value = tot;
  } else {
    node.value = flareWeightLoc(node);
  }
}

function calculateVoronoi(
  nameSoFar,
  node,
  clipPolygon,
  center,
  goodenough,
  depth
) {
  const name = nameSoFar ? `${nameSoFar}/${node.name}` : node.name;
  node.layout = {
    polygon: clipPolygon,
    center,
    algorithm: 'voronoi',
  };

  if (!node.children) {
    return;
  }
  if (depth < 3) {
    console.warn(`calculating voronoi for ${name}`);
  } else if (depth === 3) {
    console.warn(`calculating voronoi for ${name} and descendants`);
  }
  if (debug) {
    console.warn(
      `calculating voronoi for ${name} with ${node.children.length} children and a clip polygon with ${clipPolygon.length} vertices`
    );
  }

  const MAX_SIMULATION_COUNT = 200; // we re-run the whole simulation this many times if it fails
  const MAX_ITERATION_COUNT = 500; // this is how many times a particular simulation iterates
  const MIN_WEIGHT_RATIO = 0.005; // maybe this should be a parameter? Too high, we iterate a lot.  Too low, sizes are not proportional to lines of code.
  let simulationCount = 0;
  let simulationLoopEnded = false;
  let bestConvergenceRatio = 1.0;
  let bestPolygons = undefined;
  while (!simulationLoopEnded) {
    try {
      var simulation = voronoiMapSimulation(node.children)
        .maxIterationCount(MAX_ITERATION_COUNT)
        .minWeightRatio(MIN_WEIGHT_RATIO)
        .weight((d) => d.value)
        .clip(clipPolygon)
        .stop();

      var state = simulation.state();

      let tickCount = 0;
      let warningTime = Date.now();
      while (!state.ended) {
        tickCount += 1;
        const now = Date.now();
        if (now - warningTime > 10000) {
          // every 10 seconds
          warningTime = now;
          console.warn(
            `slow voronoi processing of ${name} with ${node.children.length} children, tick count: ${tickCount}`
          );
        }
        simulation.tick();
        state = simulation.state();
      }
      if (tickCount === MAX_ITERATION_COUNT) {
        if (state.convergenceRatio < bestConvergenceRatio) {
          if (debug) {
            console.warn(
              'best iteration result so far',
              simulationCount,
              state.convergenceRatio
            );
          }
          bestConvergenceRatio = state.convergenceRatio;
          bestPolygons = [...state.polygons];
        }

        if (simulationCount < MAX_SIMULATION_COUNT) {
          simulationCount = simulationCount + 1;

          console.warn(
            `processing ${name} with ${node.children.length} children - Exceeded tick count ${tickCount} - retrying from scratch, try ${simulationCount} of ${MAX_SIMULATION_COUNT}`
          );
        } else {
          console.error('Too many meta retries - stopping');
          simulationLoopEnded = true;
          if (!goodenough) {
            throw Error("Too many retries, can't provide good simulation");
          } else {
            console.warn('returning good-enough result', bestConvergenceRatio);
          }
        }
      } else {
        if (bestPolygons) {
          console.warn(
            'successful converging layout, using real ratio not best-so-far: ',
            state.convergenceRatio
          );
          bestPolygons = undefined;
          bestConvergenceRatio = state.convergenceRatio;
        }
        simulationLoopEnded = true;
      }
    } catch (e) {
      // re-try from scratch but only after predictable exceptions
      console.warn('caught e', e.message);
      if (!(e instanceof Error) && !(e instanceof TypeError)) {
        console.error('not Error or TypeError');
        throw e;
      }
      if (
        e.message === 'handleOverweighted1 is looping too much' ||
        e.message ===
          'at least 1 site has no area, which is not supposed to arise'
      ) {
        simulationCount = simulationCount + 1;
        if (simulationCount < MAX_SIMULATION_COUNT) {
          console.warn(
            `caught ${e.message}, retrying from scratch try ${simulationCount} of ${MAX_SIMULATION_COUNT}`
          );
        } else {
          console.error(
            `caught ${e.message}, too many errors!`,
            simulationCount
          );
          simulationLoopEnded = true;
          if (!goodenough) {
            throw Error("Too many retries, can't provide good simulation");
          } else {
            console.warn('returning good-enough result', bestConvergenceRatio);
          }
        }
      } else {
        console.error(
          `unhandled exception ${e.name}:${e.message} - rethrowing`
        );
        throw e;
      }
    }
  }
  var polygons = state.polygons;
  if (bestPolygons) {
    console.error(
      'No good layout found - using best convergence ratio',
      bestConvergenceRatio
    );
    polygons = bestPolygons;
  } else {
    if (debug) {
      console.warn(
        'Successful layout - best convergence ratio',
        state.convergenceRatio
      );
    }
  }

  for (const polygon of polygons) {
    const pdata = polygon.map((d) => d);
    calculateVoronoi(
      name,
      polygon.site.originalObject.data.originalData,
      pdata,
      [polygon.site.x, polygon.site.y],
      goodenough,
      depth + 1
    );
  }
}

async function read(stream) {
  const chunks = [];
  for await (const chunk of stream) chunks.push(chunk);
  return Buffer.concat(chunks).toString('utf8');
}

async function main({ input, output, points, circles, goodenough }) {
  const rawData = input
    ? await fs.readFile(input, 'utf-8')
    : await read(process.stdin);
  const width = 1024;
  const parsedData = JSON.parse(rawData);

  console.warn('getting values recursively');
  const treeData = parsedData['tree'];

  calculate_values(treeData);
  console.warn('pruning empty nodes');
  pruneWeightlessNodes(treeData);

  // top level clip shape
  if (circles) {
    // area = pi r^2 so r = sqrt(area/pi) or just use sqrt(area) for simplicity
    const children = treeData.children.map((child) => {
      return { r: Math.sqrt(child.value), originalObject: child };
    });
    d3.packSiblings(children);
    // top level layout
    const enclosingCirle = d3.packEnclose(children);
    const { x, y, r } = enclosingCirle;
    // TODO: offset by x/y
    treeData.layout = {
      polygon: computeCirclingPolygon(points, r),
      center: [0, 0],
      width: r * 2,
      height: r * 2,
      algorithm: 'circlePack',
    };

    for (const child of children) {
      const clipPolygon = computeCirclingPolygon(
        points,
        child.r
      ).map(([x, y]) => [x + child.x, y + child.y]);
      const center = [child.x, child.y];

      calculateVoronoi(
        child.originalObject.name,
        child.originalObject,
        clipPolygon,
        center,
        goodenough,
        1
      );
      child.originalObject.layout.width = child.r;
      child.originalObject.layout.height = child.r;
    }
  } else {
    const clipPolygon = computeCirclingPolygon(points, width / 2);
    const center = [0, 0];

    calculateVoronoi(null, treeData, clipPolygon, center, goodenough, 0);

    treeData.layout.width = width;
    treeData.layout.height = width;
  }

  const treeWithPaths = addPaths(null, treeData);

  parsedData['tree'] = treeWithPaths;

  console.warn('saving');
  if (output) {
    await fs.writeFile(output, JSON.stringify(parsedData));
  } else {
    process.stdout.write(JSON.stringify(parsedData));
  }
  return 'OK';
}

const argv = yargs
  //   .usage('$0 -i [input]', 'add voronoi data to json input file')
  .usage('$0 [options]')
  .alias('i', 'input')
  .describe('i', 'Input JSON file')
  .alias('o', 'output')
  .describe('o', 'Output JSON file')
  .alias('p', 'points')
  .number('p')
  .default('p', 128)
  .describe('p', 'number of points in the initial bounding circle/polygon')
  .boolean('g')
  .alias('g', 'goodenough')
  .default('g', true)
  .describe(
    'g',
    'accept a good-enough voronoi simulation, rather than failing if perfect one not found'
  )
  .describe('c', 'use circle packing for top level')
  .boolean('c')
  .alias('c', 'circles')
  .default('c', false)
  .help('h')
  .alias('h', 'help').argv;

const args = {
  input: argv.input,
  output: argv.output,
  points: argv.points,
  circles: argv.circles,
  goodenough: argv.goodenough,
};

main(args).then(
  (result) => {
    console.error('done.', result);
  },
  (err) => {
    console.error('Error thrown!', err);
    process.exit(1);
  }
);
